pkg/ottl/ottlfuncs/func_convert_case.go

```diff
@@ -0,0 +1,117 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"strings"
+	"unicode"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+func ConvertCase[K any](target ottl.Getter[K], toCase string) (ottl.ExprFunc[K], error) {
+	return func(ctx K) (interface{}, error) {
+		val, err := target.Get(ctx)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if val != nil {
+
+			if valStr, ok := val.(string); ok {
+
+				if valStr == "" {
+					return valStr, nil
+				}
+
+				switch toCase {
+				// Convert string to snake case (someName -> some_name)
+				case "snake":
```

Comment:
Lets refactor the different cases into their own unexported functions in this file.  Hopefully this will simplify things